pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,228 @@
+	r.m.Store(index, &t)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	if r.numRegistered.Load() > maxTraces {
+		return
+	}
+	r.numRegistered.Add(1)
+	// We store all the entries in the map. This is because we don't know which
+	// one was traced. Typically this is called with only a single entry.
+	t := newLocalTrace(r.ctx, baseCtx)
+	// Log the identical entry to both to make it easier to match up later.
+	t.logf(1, "registering entry (i%d) for tracing", ent.Index)
+	r.m.Store(kvpb.RaftIndex(ent.Index), &t)
+}
+
+// IndexContext is a pair of a raft index and its tracing context.
+type IndexContext struct {
+	Index uint64
+	Ctx   context.Context
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []*kvserverpb.TracedEntry {
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// Close will unregister all the traces and log them. It should be called if the
+// replica is destroyed as it is possible some traces were never closed.
+func (r *RaftTracer) Close() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.numRegistered.Add(-1)
+		r.m.Delete(index)
+		t.logf(5, "cleanup log mark (i%d) during close", index)
+		return true
+	})
+}
+
+// traceIfCovered will log the message if it is covered by a trace. Additionally
+// it returns any saved contexts by index for sending to remote nodes.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []*kvserverpb.TracedEntry {
+	var tracedEntries []*kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		switch m.Type {
+		case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+			if inRange(index, m.Entries) {
+				e := t.toTracedEntry(index)
+				tracedEntries = append(tracedEntries, &e)
+				t.logf(4, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
+			}
+		}
+		return true
+	})
+	return tracedEntries
+}
+
+// inRange returns true if the index is within the range of the entries. It
+// assumes the entries are sorted by index. Specifically it returns if the index
+// is in the set [first_index,last_index].
+func inRange(index kvpb.RaftIndex, entries []raftpb.Entry) bool {
+	return len(entries) != 0 &&
+		index >= kvpb.RaftIndex(entries[0].Index) && index <= kvpb.RaftIndex(entries[len(entries)-1].Index)
+}
+
+// traceIfPast will log the message if it is a range message the covers a trace.
```

Comment:
nit: please straighten up the grammar, e.g.:

```golang
// traceIfPast logs the event if it covers the entry index,
// for each affected entry.
```

Same for the `traceIfCovered`.

Reply:
Fixed.